File: backend/src/main/scala/order/api/CreateOrderApi.scala
package order.api

import domain.shared.given

import cats.effect.IO
import domain.order.CreateOrderRequest
import domain.shared.{DeliveryAppState, UserRole}
import order.app.createOrder
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.io.*
import shared.api.routing.*
import shared.app.*

val createOrderApi: FixedMethodApi0[DeliveryAppState] =
  jsonPostApi0[CreateOrderRequest, DeliveryAppState](
    routeSegment("api"),
    routeSegment("delivery"),
    routeSegment("orders"),
  )

File: frontend/src/apis/order/CreateOrderApi.ts
import type { CreateOrderRequest, DeliveryAppState } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { defineJsonPostApi0, routeSegment } from '@/system/api/TypedApiDefinitions'

export const createOrderApiDefinition = defineJsonPostApi0<CreateOrderRequest, DeliveryAppState>([
  routeSegment('api'),
  routeSegment('delivery'),
  routeSegment('orders'),
])

export function createOrder(payload: CreateOrderRequest) {
  return postNormalizedDeliveryState(createOrderApiDefinition, payload)
}
